src/os/ogc/ogcservice.js

```diff
@@ -0,0 +1,399 @@
+   * Get the identifying value out of this object for queries, etc
+   * @param {!T} datum
+   * @return {string|undefined|null}
+   * @template T
+   */
+  getId(datum) {
+    const id = datum.get(this.nameProperty_);
+    return id || datum.get('NAME');
+  }
+
+  /**
+   * Get the identifying value out of this object for queries, etc
+   * @param {!T} datum
+   * @return {string|undefined|null}
+   * @template T
+   */
+  getLabel(datum) {
+    const label = datum.get(this.nameProperty_);
+    return label || datum.get('NAME');
+  }
+
+  /**
+   * Get the UI
+   * @return {os.ogc.query.Query}
+   */
+  getQuery() {
+    if (!this.query_) {
+      this.query_ = new Query(this.getServiceId());
+    }
+    return this.query_;
+  }
+
+  /**
+   * Get the UI
+   * @return {string}
+   */
+  getServiceId() {
+    return this.serviceId_;
+  }
+
+  /**
+   * Get the desired tags out of the Feature
+   * @param {!ol.Feature} feature
+   * @return {undefined|string}
+   */
+  getTags(feature) {
+    var tags;
+    var props = this.properties_ ? os.settings.get(this.properties_) : null;
+    if (props) {
+      var values = [];
+      for (var i = 0; i < props.length; i++) {
+        var value = feature.get(props[i]);
+        if (value) {
+          values.push(value);
+        }
+      }
+
+      tags = values.join(', ');
+    }
+    return tags;
+  }
+
+  /**
+   * Handle area being loaded from the server.
+   * @param {!T} datum The requested area
+   * @param {!ol.Feature} feature The area border feature
+   * @return {!ol.Feature}
+   * @template T
+   */
+  populateFeature(datum, feature) {
+    var tags = this.getTags(feature);
+
+    feature.set('title', this.getLabel(feature));
+    feature.set(this.nameProperty_, undefined);
+    feature.set('tags', 'Borders' + (tags ? (', ' + tags) : ''));
+
+    this.fix_(feature);
+
+    return feature;
+  }
+
+  /**
+   * @param {!string} id
+   */
+  setServiceId(id) {
+    this.serviceId_ = id;
+  }
+
+  /**
+   * @param {function(!Array<!ol.Feature>): ?} resolve The Promise resolve function
+   * @param {function(*): ?} reject The goog.Promise reject function
+   * @private
+   */
+  loadAll_(resolve, reject) {
+    if (this.isConfigured()) {
+      var qd = os.ogc.getDefaultWfsParams();
+      qd.set('typename', this.typename_);
+      qd.set('propertyname', 'ID,' + this.nameProperty_);
+      qd.set('outputformat', 'text/xml; subtype=gml/3.1.1');
+
+      var uri = new Uri(this.url_);
+      uri.setQueryData(qd);
+
+      var request = new Request(uri);
+      request.setHeader('Accept', 'application/xml');
+      request.setValidator(os.ogc.getException);
+      request.setDataFormatter(new WFSFormatter());
+
+      request.listenOnce(goog.net.EventType.SUCCESS, goog.partial(this.onAllLoaded_, resolve, reject), false, this);
+      request.listenOnce(goog.net.EventType.ERROR, function(event) {
+        this.onError(event, 'Failed to load data list.', false, reject);
+      }.bind(this), false, this);
```

Comment:
Style thing: using `request.listen` is generally more verbose than using `request.getPromise()`. The promise version packs everything into a single call, doesn't require cleaning up listeners, and gets you directly to your result in the resolver function.

The only practical reason to use the `listen` version of a request anymore is if you need the request object again on the other side. Things like getting the response headers/response code are the cases where this is needed.

Reply:
Woot! Rewritten to use that.